### ISSUE TITLE
invalid packet header byte: 0x0009000000dd5c78

### ISSUE THREAD

**Issue by bdklahn** [Xpra-org/xpra#3470]
On the same machine (Dell Precision running OpenSUSE Tumbleweed), Xpra installed from source ("commit b1a6387b222ac3f42fd05276e61d80354601c35d (HEAD -> master, origin/master, origin/HEAD")

Konsole 1:
xpra start --start=xterm --daemon=no
(". . . OpenGL is supported on display ':2'")

Konsole 2:
xpra attach :2

server:
```
(xpra_venv) ME@HOSTNAME:~> xpra start --start=xterm --daemon=no
warning: invalid compressor(s) specified: lzo
using systemd-run to wrap 'start' server command
Running scope as unit: run-rf4090ae4c5214bc7a2539245b80f66ba.scope
failure on /tmp/.X11-unix/X18304: [Errno 2] No such file or directory: '/tmp/.X11-unix/X18304'
2022-02-19 19:06:30,821 cannot access python uinput module:
2022-02-19 19:06:30,821  No module named 'uinput'
_XSERVTransSocketUNIXCreateListener: ...SocketCreateListener() failed
_XSERVTransMakeAllCOTSServerListeners: server already running
_XSERVTransSocketUNIXCreateListener: ...SocketCreateListener() failed
_XSERVTransMakeAllCOTSServerListeners: server already running

X.Org X Server 1.21.1.3
X Protocol Version 11, Revision 0
Current Operating System: Linux HOSTNAME 5.16.5-1-default #1 SMP PREEMPT Thu Feb 3 05:26:48 UTC 2022 (1af4009) x86_64
Kernel command line: BOOT_IMAGE=/boot/vmlinuz-5.16.5-1-default root=UUID=e22794d9-0797-4e19-96d7-af07a411c29c splash=silent mitigations=auto quiet

Current version of pixman: 0.40.0
        Before reporting problems, check http://wiki.x.org
        to make sure that you have the latest version.
Markers: (--) probed, (**) from config file, (==) default setting,
        (++) from command line, (!!) notice, (II) informational,
        (WW) warning, (EE) error, (NI) not implemented, (??) unknown.
(++) Log file: "/run/user/1000/xpra/Xorg.S18304.log", Time: Sat Feb 19 19:06:30 2022
(++) Using config file: "/etc/xpra/xorg.conf"
(==) Using system config directory "/usr/share/X11/xorg.conf.d"
2022-02-19 19:06:35,402 pointer device emulation using XTest
2022-02-19 19:06:35,434 Error: cannot find the html web root
2022-02-19 19:06:35,435  '/home/ME/xpra_venv/share/xpra/www' does not exist
2022-02-19 19:06:35,435  install the html-xpra package
2022-02-19 19:06:35,452 created unix domain socket '/run/user/1000/xpra/HOSTNAME-2'
2022-02-19 19:06:35,452 cannot create group socket '/run/xpra/HOSTNAME-2'
2022-02-19 19:06:35,452  [Errno 13] Permission denied
2022-02-19 19:06:35,452 created unix domain socket '/home/ME/.xpra/HOSTNAME-2'
2022-02-19 19:06:35,732 D-Bus notification forwarding is available
2022-02-19 19:06:35,778 pulseaudio not started: 'pulseaudio' command not found
2022-02-19 19:06:35,783 removing private directory '/run/user/1000/xpra/pulse-2'
2022-02-19 19:06:35,843 watching for applications menu changes in:
2022-02-19 19:06:35,844  '/usr/share/applications'
2022-02-19 19:06:35,844  '/var/lib/snapd/desktop/applications'
2022-02-19 19:06:36,029 Warning: webcam forwarding is disabled
2022-02-19 19:06:36,030  the virtual video directory '/sys/devices/virtual/video4linux' was not found
2022-02-19 19:06:36,030  make sure that the 'v4l2loopback' kernel module is installed and loaded
2022-02-19 19:06:36,030  or use the 'webcam=no' option
2022-02-19 19:06:36,031 found 0 virtual video devices for webcam forwarding

(Xpra:18304): Gtk-CRITICAL **: 19:06:36.153: gtk_widget_realize: assertion 'widget->priv->anchored || GTK_IS_INVISIBLE (widget)' failed
2022-02-19 19:06:36,269 GStreamer version 1.18.6 for Python 3.8.12 64-bit
2022-02-19 19:06:36,364 xpra is ready.
2022-02-19 19:06:37,587 loaded 88 start menu entries from 11 sub-menus in 1.7 seconds
2022-02-19 19:06:38,493 xpra X11 seamless version 4.2.3 64-bit
2022-02-19 19:06:38,494  uid=1000 (ME), gid=100 (users)
2022-02-19 19:06:38,494  running with pid 18304 on Linux unknown unknown unknown
2022-02-19 19:06:38,496  connected to X11 display :2 with 24 bit colors
2022-02-19 19:06:39,326 printer forwarding enabled using postscript and pdf
2022-02-19 19:06:39,343 started command 'xterm' with pid 18394
2022-02-19 19:06:39,446 62.5GB of system memory
2022-02-19 19:06:40,207 OpenGL is supported on display ':2'
2022-02-19 19:06:40,208  using 'llvmpipe (LLVM 13.0.1, 256 bits)' renderer
2022-02-19 19:15:54,744 New unix-domain connection received
2022-02-19 19:15:54,744  on '/run/user/1000/xpra/HOSTNAME-2'
2022-02-19 19:15:54,750 Disconnecting client /run/user/1000/xpra/HOSTNAME-2:
2023-02-19 19:15:54,751  invalid packet header byte: 0x0009000000dd5c78 read buffer=b'\x00\t\x00\x00\x00\xdd\\x\xda\xcc\xbdY\xb3\xe3 .. d\xad\xa0\xab\xd6\x81f\x01\x01\x00\x10\xfdC\xec' (56675 bytes)

```

client:
```
(xpra_venv) ME@HOSTNAME:~/git_repos/xpra> xpra attach :2
warning: invalid compressor(s) specified: lzo
2022-02-19 19:15:50,473 Xpra GTK3 X11 client version 4.4-r28472 (g2f8bf9fcb) 64-bit
2022-02-19 19:15:50,511  running on Linux unknown unknown unknown
2022-02-19 19:15:50,513  window manager is 'KWin'
2022-02-19 19:15:51,689 GStreamer version 1.18.6 for Python 3.8.12 64-bit
2022-02-19 19:15:51,873 created unix domain socket '/run/user/1000/xpra/clients/HOSTNAME-18844'
2022-02-19 19:15:52,704 OpenGL_accelerate module loaded
2022-02-19 19:15:52,709 Using accelerated ArrayDatatype
2022-02-19 19:15:52,878 Warning: cannot handle window transparency
2022-02-19 19:15:52,878  screen is not composited
2022-02-19 19:15:52,895 OpenGL enabled with NV136
2022-02-19 19:15:53,128  keyboard settings: rules=evdev, model=pc105, layout=us
2022-02-19 19:15:54,143  desktop size is 3840x2160:
2022-02-19 19:15:54,144   :0.0 (1016x571 mm - DPI: 96x96) workarea: 3840x2114
2022-02-19 19:15:54,145     DEL DP-1         (609x349 mm - DPI: 160x157)
2022-02-19 19:15:54,753 Error: failed to connect
2022-02-19 19:15:54,753  received uninterpretable nonsense: 'invalid packet header byte: 0x646973636f6e6e65'
2022-02-19 19:15:54,753  b"disconnect: invalid packet header byte: 0x0009 .. \x01\\x01\\x00\\x10\\xfdC\\xec' (56675 bytes)\n" (186 bytes)
2022-02-19 19:15:54,753  data: "disconnect: invalid packet header byte: 0x0009000000dd5c78 read buffer=b'\\x00\\t\\x00\\x00\\x00\\xdd\\\\x\\xda\\xcc\\xbdY\\xb3\\xe3 .. d\\xad\\xa0\\xab\\xd6\\x81f\\x01\\x01\\x00\\x10\\xfdC\\xec' (56675 bytes)"
2022-02-19 19:15:54,793 removing unix domain socket '/run/user/1000/xpra/clients/HOSTNAME-18844'
```

Not sure why it complains about 'uinput', because it is installed and that module is loaded in the kernel (I can import it in python interpreter and lsmod shows uinput).

Below is an unedited concise accounting of what I've done to enable it to build.
Some things were done and undone. Yes, I know that not all the things were necessary nor critical, but I wanted to rule out things by squashing even warnings (and maybe enabling all the capabilities).

```
pip install cython
Package 'xres', required by 'virtual:world', not found
sudo zypper install libXres-devel
Package 'pygobject-3.0', required by 'virtual:world', not found
pip install pygobject
Package 'pygobject-3.0', required by 'virtual:world', not found
pip uninstall pygobject
sudo zypper install python38-gobject-devel
ModuleNotFoundError: No module named 'gi'
pip install gi-docgen
ModuleNotFoundError: No module named 'gi'
pip uninstall gi-docgen
sudo zypper install python3-gi-docgen
ModuleNotFoundError: No module named 'gi'
sudo zypper remove python3-gi-docgen
# from https://pygobject.readthedocs.io/en/latest/getting_started.html
pip3 install PyGObject
ModuleNotFoundError: No module named 'PIL
pip install pillow
Error: cannot enable SSH socket upgrades:
No module named 'paramiko'
pip install paramiko
Error setting up dbus signals:
 No module named 'dbus'
pip install dbus-python
Warning: cannot import native OpenGL module
 No module named 'OpenGL'
pip install pyopengl
No OpenGL_accelerate module loaded: No module named 'OpenGL_accelerate'
pip install pyopengl-accelerate
Warning: printer forwarding disabled:
 No module named 'cups'
pip install pycups
cupsconnection.h:26:10: fatal error: cups/http.h: No such file or directory
sudo zypper install cups-devel
pip install pycups
Warning: cannot use application menu data:
 no python-xdg module
ModuleNotFoundError: No module named 'xdg'
pip install xdg
sudo zypper install python38-xdg
Warning: cannot use application menu data:
 no python-xdg module
ModuleNotFoundError: No module named 'xdg.DesktopEntry'
sudo zypper install python38-pyxdg
cannot access python uinput module:
 No module named 'uinput'
pip install python-uinput
```

**Comment by totaam** [comment]
> `warning: invalid compressor(s) specified: lzo`

Your `/etc/xpra` config files are from an old version

> `cannot access python uinput module`

You do not have the permissions required for opening `/dev/uinput` using `uinput.fdopen()`.
Unless you have very specific needs when it comes to input devices, don't bother with `uinput` - which can cause more problems than it is worth.

> `cannot create group socket '/run/xpra/HOSTNAME-2'`

You are not a member of the `xpra` group or `/run/xpra` is not created properly (see https://github.com/Xpra-org/xpra/blob/master/fs/lib/tmpfiles.d/xpra.conf)
Not really a problem unless you intend to share sockets with other users.

> `pulseaudio not started: 'pulseaudio' command not found`

Audio forwarding will not work, or it may forward audio from the main session instead.

> `Xpra GTK3 X11 client version 4.4-r28472`

If the revision value is correct, that's very old. Over 1500 revisions old.

> ` window manager is 'KWin'`
> (..)
> `Warning: cannot handle window transparency`
> ` screen is not composited`

I thought `KWin` was a compositing window manager?

The header byte issue already has a ticket and the fix has been released: #3346

---

> Below is an unedited concise accounting of what I've done to enable it to build.

You should never install via `pip` when your package manager can provide all these packages.

**Comment by bdklahn** [comment]
> > `warning: invalid compressor(s) specified: lzo`
> 
> Your `/etc/xpra` config files are from an old version
I did a `sudo rm -rf /etc/xpra` before even starting.
Where do you think Xpra is pulling old config files from?
> 
> > `cannot access python uinput module`
> 
> You do not have the permissions required for opening `/dev/uinput` using `uinput.fdopen()`. Unless you have very specific needs when it comes to input devices, don't bother with `uinput` - which can cause more problems than it is worth.
Ok. Not critical. Thanks.
But a worse than useless message: It's misleading. Anything that can be done about that here?
> 
> > `cannot create group socket '/run/xpra/HOSTNAME-2'`
> 
> You are not a member of the `xpra` group or `/run/xpra` is not created properly (see https://github.com/Xpra-org/xpra/blob/master/fs/lib/tmpfiles.d/xpra.conf) Not really a problem unless you intend to share sockets with other users.
I did see something in the docs about that being non-critical, etc.
> 
> > `pulseaudio not started: 'pulseaudio' command not found`
> 
> Audio forwarding will not work, or it may forward audio from the main session instead.
Thanks. I thought I had pulseaudio installed, but maybe not.
> 
> > `Xpra GTK3 X11 client version 4.4-r28472`
> 
> If the revision value is correct, that's very old. Over 1500 revisions old.
Maybe GTK is not a critical thing on this KDE Qt system???
Not sure what to do about this (especially if you think I should only use system packages). I would assume a rolling release (Tumbleweed), updated yesterday, would have as new as possible.
> 
> > ` window manager is 'KWin'`
> > (..)
> > `Warning: cannot handle window transparency`
> > ` screen is not composited`
> 
> I thought `KWin` was a compositing window manager?
I would have thought so, as well.
> 
> The header byte issue already has a ticket and the fix has been released: #3346
I saw Issue #3346. I assumed that would already be in the HEAD of master, which I'm building from (see the first sentence in this Issue). I would consider that at "hotfix"-type issue. Has that not made it into master?
> 
> > Below is an unedited concise accounting of what I've done to enable it to build.
> 
> You should never install via `pip` when your package manager can provide all these packages.
I did that to:
1. Use venv to provide some isolation from my workstation (short of just running in a clean VM install).
2. Have presumably newer versions of things (with more bug fixes, etc.). -to better isolate Xpra-specific bugs.

I guess I'm a little confused. The best way to install Xpra is to **not** use system packages, but the best way to install everything else is to use system packages??? I understand not mixing (thus I uninstalled the system Xpra package before starting this). But, given those Python packages must not have installed by the system, I'm not sure the harm in installing them in a venv.

**Comment by bdklahn** [comment]
Is this the fix to that header byte Issue #3346?
https://github.com/Xpra-org/xpra/blob/d026c1ed6d1a99ac7560ffad8a545966dcdf4017/xpra/server/server_core.py#L1336

That seems to be included in the snapshot I am building from.
And I am running from . . .
```
(xpra_venv) . . . ~/git_repos/xpra> type xpra
xpra is hashed (/home/<me>/xpra_venv/bin/xpra)
```

**Comment by bdklahn** [comment]
Ok. Now I am more confused.
I looked at my system packages, and xpra was shown to be installed there.

So I uninstalled it, made sure /etc/xpra was removed.
Then I went into my repo, with the updated master pulled, and my xpra_venv activated, and did . . .
`python3 ./setup.py install`

and . . .
```
(xpra_venv) <me>@<hostname>:~/git_repos/xpra> type xpra
xpra is hashed (/home/<me>/xpra_venv/bin/xpra)
```

But then . . .
```
~/git_repos/xpra> xpra start --start=xterm --daemon=no
using systemd-run to wrap 'start' server command
Running scope as unit: run-ree7bdde4377c43ee9adf65ea1bbab360.scope
Traceback (most recent call last):
  File "/home/<me>/xpra_venv/bin/xpra", line 5, in <module>
    from xpra.platform import init, set_default_name
ModuleNotFoundError: No module named 'xpra'
```

Can't Xpra's setup.py make sure and install it's own direct Python dependencies in a venv?

**Comment by bdklahn** [comment]
Ok. So the OpenSUSE package maintainer responded, saying he doesn't use Xpra and doesn't have time to continue maintaining the SUSE package. He (almost predictably) asked if I wanted to take over.
I could elaborate, but if you have read access to their Bugzilla: [https://bugzilla.suse.com/show_bug.cgi?id=1195873](https://bugzilla.suse.com/show_bug.cgi?id=1195873)

I'm considering more contained/reproduce able run testing strategies (Docker, or qemu KVM on custom KIWI NG snapshots, etc.), I also asked about more automated post-build testing capabilities of OBS, etc.

But this, on the main OpenSUSE page, also caught my eye.
[open.qa](http://open.qa/)
Looks interesting. Anyway, I'm of course hoping to make things simpler for me to run on OpenSUSE. But I also hope, maybe, some of this could contribute to any streamlining here.
But maybe you already know about that. I seem to recall our devs bringing that up a couple years ago, and then there being an issue where it would end up being more work to maintain. Idk. Just thoughts.

**Comment by bdklahn** [comment]
Sigh. Ok. I did a fresh install of OpenSUSE Tumbleweed KDE in Virtual Machine Manager (libvirt qemu KVM). Then 'sudo zypper install xpra'. It installed, like, 37 packages.

Then it just worked.

Then I went back and did the same on my regular machine.
That still didn't work.
I checked that /etc/xpra was gone, first.
After I ran xpra on the problematic machine, /etc/xpra appeared. It seemed to have the same files and timestamps as the working install in the VM.

Is there another place I might look to find old files or setting which might be messing up the xpra install/update?

**Comment by bdklahn** [comment]
Just thought I'd try and remove everything with rpm -e

```
#repo purge 
#!/bin/sh
for package in "$@"
do  
  echo "removing config files for $package"  
  for file in $(rpm -q --configfiles $package)  
  do    
    echo "  removing $file"    
    rm -f $file  
  done  
rpm -e $package
done
```
See [here](https://together.jolla.com/question/229207/zypper-remove-packages-including-configs-apt-purge/).

Then . . .

```
> sudo ./repo_purge.sh xpra
[sudo] password for root: 
removing config files for xpra
  removing /etc/X11/xorg.conf.d/90-xpra-virtual.conf
  removing /etc/dbus-1/system.d/xpra.conf
  removing /etc/pam.d/xpra
  removing /etc/xpra/conf.d/05_features.conf
  removing /etc/xpra/conf.d/10_network.conf
  removing /etc/xpra/conf.d/12_ssl.conf
  removing /etc/xpra/conf.d/15_file_transfers.conf
  removing /etc/xpra/conf.d/16_printing.conf
  removing /etc/xpra/conf.d/20_sound.conf
  removing /etc/xpra/conf.d/30_picture.conf
  removing /etc/xpra/conf.d/35_webcam.conf
  removing /etc/xpra/conf.d/40_client.conf
  removing /etc/xpra/conf.d/42_client_keyboard.conf
  removing /etc/xpra/conf.d/50_server_network.conf
  removing /etc/xpra/conf.d/55_server_x11.conf
  removing /etc/xpra/conf.d/60_server.conf
  removing /etc/xpra/conf.d/65_proxy.conf
  removing /etc/xpra/content-categories/10_default.conf
  removing /etc/xpra/content-type/10_role.conf
  removing /etc/xpra/content-type/30_title.conf
  removing /etc/xpra/content-type/50_class.conf
  removing /etc/xpra/content-type/70_commands.conf
  removing /etc/xpra/http-headers/00_nocache.txt
  removing /etc/xpra/http-headers/10_content_security_policy.txt
  removing /etc/xpra/xorg-uinput.conf
  removing /etc/xpra/xorg.conf
  removing /etc/xpra/xpra.conf
warning: file /etc/xpra/xpra.conf: remove failed: No such file or directory
warning: file /etc/xpra/xorg.conf: remove failed: No such file or directory
warning: file /etc/xpra/xorg-uinput.conf: remove failed: No such file or directory
warning: file /etc/xpra/http-headers/10_content_security_policy.txt: remove failed: No such file or directory
warning: file /etc/xpra/http-headers/00_nocache.txt: remove failed: No such file or directory
warning: file /etc/xpra/content-type/70_commands.conf: remove failed: No such file or directory
warning: file /etc/xpra/content-type/50_class.conf: remove failed: No such file or directory
warning: file /etc/xpra/content-type/30_title.conf: remove failed: No such file or directory
warning: file /etc/xpra/content-type/10_role.conf: remove failed: No such file or directory
warning: file /etc/xpra/content-categories/10_default.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/65_proxy.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/60_server.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/55_server_x11.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/50_server_network.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/42_client_keyboard.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/40_client.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/35_webcam.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/30_picture.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/20_sound.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/16_printing.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/15_file_transfers.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/12_ssl.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/10_network.conf: remove failed: No such file or directory
warning: file /etc/xpra/conf.d/05_features.conf: remove failed: No such file or directory
warning: file /etc/pam.d/xpra: remove failed: No such file or directory
warning: file /etc/dbus-1/system.d/xpra.conf: remove failed: No such file or directory
warning: file /etc/X11/xorg.conf.d/90-xpra-virtual.conf: remove failed: No such file or directory
```

Still doesn't work on the non-VM install.

**Comment by totaam** [comment]
Sorry, but I have no idea.
Something is different between your two systems... divide and conquer until you narrow it down enough.

**Comment by bdklahn** [comment]
The difference appears to be the machine.
1724025f0690a6aaf517a10dd8eb6bc243b6cc67 appears to have not made it into the OpenSUSE package, yet.

But it only seems to cause the problem on my two Dell machines (Precision and Optiplex).
That bug is still present on my Lenovo Yoga, and in the Qemu VM running on  Precision. But it doesn't cause problems on those last two.
Am I right that there was a race condition, with that bug? Maybe my/the Dell network hardware reveals the bug, but not so in the Lenovo. And maybe the virtualization in the VM limits things there, where the condition is not met, as well.

I did a __complete__ reinstall on my Precision, and was surprised to see the bug return (despite those newer/cleaner installations working). I checked /usr/lib64/python3.8/site-packages/xpra/server/server_core.py, and that problematic underscore was in there. I added the underscore, and then it worked. I also checked that the bug was still in the (working) Lenovo OS.

I guess installing from source, and running, in a venv, doesn't help. It seems to end up falling back to the (broken) system install. I guess it must jump out of/ignore the venv if/when system paths get called and/or the systemd stuff might get invoked?

So I guess this can be closed.
